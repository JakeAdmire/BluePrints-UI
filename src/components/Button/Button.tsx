import React, { Component, MouseEvent } from "react";
// 
import * as helpers from '../../misc/helpers';
import Color from 'color';

//-------------------------| Button Varients |-------------------------//
import { DefaultButton } from './varients/DefaultButton';
import { NeumorphismButton } from './varients/NeumorphismButton';
    // toggleable
import { ToggleableDefaultButton } from "./varients/ToggleableDefaultButton";
import { ToggleableNeumorphismButton } from './varients/ToggleableNeumorphismButton';


export type ButtonProps = {
    url?: string,
    text?: string,
    overrideStyles?: React.CSSProperties,
    textColor?: string,
    onClick?: (event: MouseEvent<HTMLButtonElement | HTMLLabelElement>) => void
} & Partial<DefaultProps>;

type DefaultProps = {
    variant: "default" | "neumorphism",
    toggleable: boolean,
    backgroundColor: string
}

type ButtonState = {
    child: string,
    textColor: string,
    lowlight: string,
    highlight: string
}

export class Button extends Component<ButtonProps, ButtonState> {
    constructor(props: ButtonProps) {
        super(props);
        this.state = {
            child: 'DefaultButton',
            textColor: this.props.textColor,
            lowlight: null,
            highlight: null
        }
    }
    static defaultProps: DefaultProps = {
        variant: "default",
        backgroundColor: "#0299E3",
        toggleable: false
    }

    componentDidMount() {
        const newChild = this.determineChild();
        this.buildAdditionalProps(newChild);
    }

    determineChild(): string {
        const { toggleable, variant } = this.props;
        const child: string = (toggleable ? 'Toggleable' : '') + helpers.toTitleCase(variant) + 'Button';

        this.setState({ child });
        
        return child;
    }

    determineTextColor(backgroundColor: string): string {
        return Color(backgroundColor).isDark()
            ? '#FFF'
            : '#000';
    }

    buildAdditionalProps(newChild: string): void {
        if (newChild.includes('Neumorphism')) this.calculateLighting();
    }

    calculateLighting(): void {
        const backgroundColor = Color(this.props.backgroundColor);

        this.setState({
            lowlight: backgroundColor.darken(.4).alpha(.15).string(),
            highlight: backgroundColor.lighten(.4).alpha(.15).string()
        })
    }

    gatherProps(): any {
        const initialProps = this.props;
        const { child, textColor, lowlight, highlight } = this.state;
        const determineTextColor = this.determineTextColor;

        if (child.includes("Neumorphism"))
            return {...initialProps, textColor, lowlight, highlight};

        else return {...initialProps, textColor, determineTextColor};
    }

    render() {
        const buttonProps = this.gatherProps();

        switch (this.state.child) {
            case "NeumorphismButton":
                return <NeumorphismButton {...buttonProps} />
            
            case "ToggleableNeumorphismButton":
                return <ToggleableNeumorphismButton {...buttonProps} />

            case "ToggleableDefaultButton":
                return <ToggleableDefaultButton {...buttonProps} />

            default:
                return <DefaultButton {...buttonProps} />;
        }
    }
}
