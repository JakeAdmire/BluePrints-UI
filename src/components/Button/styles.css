.generalButton {
    box-sizing: border-box;
    display: inline-block;
    position: relative;
    padding: .5rem 1rem;
    outline: none;
    border-radius: 10px;
}

.generalButton p {
    margin: 0;
    font-size: 0.875rem;
    font-family: "Roboto", "Helvetica", "Arial", sans-serif;
    font-weight: 500;
    line-height: 1.75;
    letter-spacing: 0.02857em;
    text-transform: uppercase;
    user-select: none;
}

.generalButton:hover {
    cursor: pointer;
}
