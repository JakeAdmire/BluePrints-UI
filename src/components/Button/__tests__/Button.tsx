import React from "react";
import { configure, shallow, ShallowWrapper } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import { Button } from '../Button';


configure({ adapter: new Adapter() });
describe('Button', () => {
    let wrapper: ShallowWrapper;

    beforeEach(() => {
        wrapper = shallow(<Button />);
    })

    it('matches the snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    })

    it('should have default props', () => {
        const wrapperProps = wrapper.props();
        const defaultProps = {
            backgroundColor: "#7d7d7d",
            disabled: false,
            text: "default",
            toggleable: false,
            variant: "default",
        };

        expect(wrapperProps['backgroundColor']).toEqual(defaultProps.backgroundColor);
        expect(wrapperProps['disabled']).toEqual(defaultProps.disabled);
        expect(wrapperProps['text']).toEqual(defaultProps.text);
        expect(wrapperProps['toggleable']).toEqual(defaultProps.toggleable);
        expect(wrapperProps['variant']).toEqual(defaultProps.variant);
    })

    it('should have default state', () => {
        const wrapperState = wrapper.state();

        type ButtonState = {
            child: string,
            highlight: string,
            lowlight: string,
            textColor: string,
        };
        const defaultState: ButtonState = {
            child: 'DefaultButton',
            textColor: undefined,
            lowlight: null,
            highlight: null
        };

        expect(wrapperState['child']).toEqual(defaultState.child);
        expect(wrapperState['textColor']).toEqual(defaultState.textColor);
        expect(wrapperState['lowlight']).toEqual(defaultState.lowlight);
        expect(wrapperState['highlight']).toEqual(defaultState.highlight);
    })

    describe.skip('On Mount...', () => {

        it('should invoke checkForErrors', () => {
    
        })

        describe('checkForErrors()', () => {

            it('should throw an error if non-compatible props are used', () => {
    
            })

        })

        it('should invoke determineChild', () => {

        })

        describe('determineChild()', () => {

            it('should return a string', () => {

            })

        })

        it('should invoke buildAdditionalProps', () => {

        })

        describe('buildAdditionalProps()', () => {

            it('should invoke calculateLighting if the component is a neumorphic variant', () => {

            })

        })

    })

    describe.skip('On Render...', () => {

        it('should render DefaultButton by default', () => {

        })

        it('should render the appropriate element depending on the given props', () => {
            
        })

        it('should invoke gatherProps', () => {

        })

        describe('gatherProps()', () => {

            it('should return a prop object', () => {

            })

            it('should invoke determineTextColor', () => {

            })

            describe('determineTextColor()', () => {

            })

        })

    })

})