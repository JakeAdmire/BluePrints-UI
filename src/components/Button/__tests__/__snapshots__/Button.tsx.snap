// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button matches the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Button
    backgroundColor="#7d7d7d"
    disabled={false}
    text="default"
    toggleable={false}
    variant="default"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "backgroundColor": "#7d7d7d",
      "determineTextColor": [Function],
      "disabled": false,
      "text": "default",
      "textColor": undefined,
      "toggleable": false,
      "variant": "default",
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "backgroundColor": "#7d7d7d",
        "determineTextColor": [Function],
        "disabled": false,
        "text": "default",
        "textColor": undefined,
        "toggleable": false,
        "variant": "default",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
